src/prommis/conn/connectivity.py

```diff
@@ -0,0 +1,581 @@
+                f.write("</script></body></html>")
+            case _:
+                raise ValueError(f"Bad output format: {output_format}")
+        if output_file is None:
+            return f.getvalue()
+
+    def _body(self, outfile):
+        i = self.indent
+        outfile.write("flowchart TD\n")
+        # Get connections first, so we know which streams to show
+        connections, show_streams = self._get_connections()
+        # Units
+        for s in self._get_mermaid_units():
+            outfile.write(f"{i}{s}\n")
+        # Streams
+        for abbr, s in self._get_mermaid_streams():
+            if abbr in show_streams:
+                outfile.write(f"{i}{s}\n")
+        # Connections
+        for s in connections:
+            outfile.write(f"{i}{s}\n")
+
+    def _get_mermaid_units(self):
+        for name, abbr in self._conn.units.items():
+            yield f"{abbr}[{name}]"
+
+    def _get_mermaid_streams(self):
+        for name, abbr in self._conn.streams.items():
+            yield abbr, f"{abbr}([{name}])"
+
+    def _get_connections(self):
+        connections = []
+        show_streams = set()
+        for stream_abbr, values in self._conn.connections.items():
+            if values[0] is not None and values[1] is not None:
+                connections.append(f"{values[0]} --> {values[1]}")
+            elif values[0] is not None:
+                connections.append(f"{values[0]} --> {stream_abbr}")
+                show_streams.add(stream_abbr)
+            elif values[1] is not None:
+                connections.append(f" {stream_abbr} --> {values[1]}")
+                show_streams.add(stream_abbr)
+        return connections, show_streams
+
+
+@dataclass
+class Connectivity:
+    units: dict = field(default_factory=dict)
+    streams: dict = field(default_factory=dict)
+    connections: dict = field(default_factory=dict)
+
+
+class ConnectivityFromFile:
+    """Build connectivity information from input data."""
+
+    def __init__(self, input_file: str | Path | TextIO):
+        if isinstance(input_file, str) or isinstance(input_file, Path):
+            datafile = open(input_file, "r")
+        else:
+            datafile = input_file
+        reader = csv.reader(datafile)
+        self._header = next(reader)
+        self._rows = list(reader)
+        self._c = None
+
+    @property
+    def connectivity(self):
+        if self._c is None:
+            units = self._build_units()
+            streams = self._build_streams()
+            connections = self._build_connections(units, streams)
+            self._c = Connectivity(
+                units=units, streams=streams, connections=connections
+            )
+        return self._c
+
+    def _build_units(self):
+        units = {}
+        c1, c2 = 1, -1
+        for s in self._header[1:]:
+            abbr = "Unit_"
+            if c2 > -1:
+                abbr += chr(ord("A") + c2)
+            abbr += chr(ord("A") + c1)

```

Comment:
I don't know what `chr()` or `ord()` do off the top of my head, so I'd suggest to add a comment to clarify.

Reply:
ah kids today. ok